Incorrect FIRRTL for BoringUtils upwards read

If you use `BoringUtils.tap` or `BoringUtils.tapAndRead` that results in an upwards reference, this will create an input port that does not have flips stripped.  This results in illegal FIRRTL and will error out in CIRCT.

Example:

```scala
//> using scala "2.13.11"
//> using lib "org.chipsalliance::chisel::6.0.0-M3"
//> using plugin "org.chipsalliance:::chisel-plugin::6.0.0-M3"
//> using options "-unchecked", "-deprecation", "-language:reflectiveCalls", "-feature", "-Xcheckinit", "-Xfatal-warnings", "-Ywarn-dead-code", "-Ywarn-unused", "-Ymacro-annotations"

import chisel3._
import chisel3.util.{Decoupled, DecoupledIO}
import chisel3.util.experimental.BoringUtils
import circt.stage.ChiselStage

class Bar(b: DecoupledIO[Bool]) extends RawModule {
  BoringUtils.tapAndRead(b)
}

class Foo extends RawModule {
  val a = WireInit(Decoupled(Bool()), DontCare)
  dontTouch(a)

  val bar = Module(new Bar(a))
}

object Main extends App {
  println(
    ChiselStage.emitCHIRRTL(new Foo)
  )
  println(
    ChiselStage.emitSystemVerilog(
      gen = new Foo,
      firtoolOpts = Array("-disable-all-randomization", "-strip-debug-info")
    )
  )
}
```

This produces (`scala-cli Foo.scala`) the following CHIRRTL:

```
FIRRTL version 3.1.0
circuit Foo :%[[
  {
    "class":"firrtl.transforms.DontTouchAnnotation",
    "target":"~Foo|Foo>a"
  }
]]
  module Bar :
    input bore : { flip ready : UInt<1>, valid : UInt<1>, bits : UInt<1>}snippets/BoringUtils.scala 12:25]

    skip

  module Foo :

    wire a : { flip ready : UInt<1>, valid : UInt<1>, bits : UInt<1>}snippets/BoringUtils.scala 16:19]
    invalidate a.bits
    invalidate a.valid
    invalidate a.ready
    inst bar of Bar
    connect bar.bore, a
```

The error produced during FIRRTL compilation is:

```
Exception in thread "main" circt.stage.phases.Exceptions$FirtoolNonZeroExitCode: firtool returned a non-zero exit code. Note that this version of Chisel (6.0.0-M3) was published against firtool version 1.52.0.
------------------------------------------------------------------------------
ExitCode:
1
STDOUT:

STDERR:
<unknown>:0: warning: option -dedup is deprecated since firtool 1.57.0, has no effect (deduplication is always enabled), and will be removed in firtool 1.58.0
Users/schuylereldridge/repos/github.com/seldridge/scala-snippets/BoringUtils.scala:12:25: error: port "bore_ready" not fully initialized in module "Bar"
```

--- a/src/main/scala/chisel3/util/experimental/BoringUtils.scala
+++ b/src/main/scala/chisel3/util/experimental/BoringUtils.scala
@@ -220,7 +220,9 @@ object BoringUtils {
   private def boreOrTap[A <: Data](source: A, createProbe: Option[ProbeInfo] = None)(implicit si: SourceInfo): A = {
     import reflect.DataMirror
     def parent(d: Data): BaseModule = d.topBinding.location.get
-    def purePortTypeBase = if (DataMirror.hasOuterFlip(source)) Flipped(chiselTypeOf(source)) else chiselTypeOf(source)
+    def purePortTypeBase = if (createProbe.nonEmpty) Output(chiselTypeOf(source))
+    else if (DataMirror.hasOuterFlip(source)) Flipped(chiselTypeOf(source))
+    else chiselTypeOf(source)
     def purePortType = createProbe match {
       case Some(pi) if pi.writable => RWProbe(purePortTypeBase)
       case Some(pi)                => Probe(purePortTypeBase)

--- a/src/test/scala/chiselTests/BoringUtilsTapSpec.scala
+++ b/src/test/scala/chiselTests/BoringUtilsTapSpec.scala
@@ -472,7 +472,6 @@ class BoringUtilsTapSpec extends ChiselFlatSpec with ChiselRunners with Utils wi
       val a = WireInit(DecoupledIO(Bool()), DontCare)
       val dummyA = Wire(Output(chiselTypeOf(a)))
       // FIXME we shouldn't need this intermediate wire
-      // https://github.com/chipsalliance/chisel/issues/3556
       // https://github.com/chipsalliance/chisel/issues/3557
       dummyA :#= a
       dontTouch(a)

--- a/src/test/scala/chiselTests/BoringUtilsTapSpec.scala
+++ b/src/test/scala/chiselTests/BoringUtilsTapSpec.scala
@@ -425,4 +425,75 @@ class BoringUtilsTapSpec extends ChiselFlatSpec with ChiselRunners with Utils wi
     // Check that firtool also passes
     val verilog = circt.stage.ChiselStage.emitSystemVerilog(new Top(Definition(new Foo)))
   }
+
+  it should "work with DecoupledIO in a hierarchy" in {
+    import chisel3.util.{Decoupled, DecoupledIO}
+    class Bar() extends RawModule {
+      val decoupledThing = Wire(Decoupled(Bool()))
+      decoupledThing := DontCare
+    }
+    class Foo() extends RawModule {
+      val bar = Module(new Bar())
+    }
+    class FakeView(foo: Foo) extends RawModule {
+      val decoupledThing = Wire(DecoupledIO(Bool()))
+      decoupledThing := BoringUtils.tapAndRead(foo.bar.decoupledThing)
+    }
+    class Top() extends RawModule {
+      val foo = Module(new Foo())
+      val fakeView = Module(new FakeView(foo))
+    }
+
+    val chirrtl = circt.stage.ChiselStage.emitCHIRRTL(new Top(), Array("--full-stacktrace"))
+    matchesAndOmits(chirrtl)(
+      "module Bar :",
+      "output decoupledThing_bore : Probe<{ ready : UInt<1>, valid : UInt<1>, bits : UInt<1>}>",
+      "define decoupledThing_bore = probe(decoupledThing)",
+      "module Foo :",
+      "output decoupledThing_bore : Probe<{ ready : UInt<1>, valid : UInt<1>, bits : UInt<1>}>",
+      "define decoupledThing_bore = bar.decoupledThing_bore",
+      "module FakeView :",
+      "input decoupledThing_bore : { ready : UInt<1>, valid : UInt<1>, bits : UInt<1>}",
+      "module Top :",
+      "connect fakeView.decoupledThing_bore, read(foo.decoupledThing_bore)"
+    )()
+
+    // Check that firtool also passes
+    val verilog = circt.stage.ChiselStage.emitSystemVerilog(new Top())
+  }
+
+  it should "work with DecoupledIO" in {
+    import chisel3.util.{Decoupled, DecoupledIO}
+    class Bar(b: DecoupledIO[Bool]) extends RawModule {
+      BoringUtils.tapAndRead(b)
+    }
+
+    class Foo extends RawModule {
+      val a = WireInit(DecoupledIO(Bool()), DontCare)
+      val dummyA = Wire(Output(chiselTypeOf(a)))
+      // FIXME we shouldn't need this intermediate wire
+      // https://github.com/chipsalliance/chisel/issues/3556
+      // https://github.com/chipsalliance/chisel/issues/3557
+      dummyA :#= a
+      dontTouch(a)
+
+      val bar = Module(new Bar(dummyA))
+    }
+
+    val chirrtl = circt.stage.ChiselStage.emitCHIRRTL(new Foo, Array("--full-stacktrace"))
+
+    matchesAndOmits(chirrtl)(
+      "module Bar :",
+      "input bore : { ready : UInt<1>, valid : UInt<1>, bits : UInt<1>}",
+      "module Foo :",
+      "wire a : { flip ready : UInt<1>, valid : UInt<1>, bits : UInt<1>}",
+      // FIXME shouldn't need intermediate wire
+      "wire dummyA : { ready : UInt<1>, valid : UInt<1>, bits : UInt<1>}",
+      "connect bar.bore, dummyA"
+    )()
+
+    // Check that firtool also passes
+    val verilog = circt.stage.ChiselStage.emitSystemVerilog(new Foo)
+  }
+
 }

